### ISSUE TITLE
Schema migrations fail when upgrading från 1.8.1 to 1.9.0

### ISSUE THREAD

**Issue by maggu** [Icinga/icingaweb2-module-director#2482]
Applying schema migrations in the GUI, after upgrading från 1.8.1 to 1.9.0, results in the following error:

> Migration 174 failed (SQLSTATE[42701]: Duplicate column: 7 ERROR: column "uuid" of relation "icinga_zone" already exists) while running ALTER TABLE icinga_zone ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16)

## Your Environment

* Director version (System - About): 1.9.0
* Icinga Web 2 version and modules (System - About): 2.9.5
* Icinga 2 version (`icinga2 --version`): 2.13.2-1
* Operating System and version: CentOS Linux release 7.9.2009 (Core)
* Webserver, PHP versions: Apache HTTP Server 2.4.6, PHP 7.3.29
* **Database**: PostgreSQL server 9.2.24

**Comment by Thomas-Gelf** [comment]
Please share the output of:

\d icinga_zone

...and:

SELECT * FROM director_schema_migration;

**Comment by maggu** [comment]
```
director=> \d icinga_zone
                                   Table "public.icinga_zone"
   Column    |          Type          |                        Modifiers
-------------+------------------------+----------------------------------------------------------
 id          | integer                | not null default nextval('icinga_zone_id_seq'::regclass)
 parent_id   | integer                |
 object_name | character varying(255) | not null
 object_type | enum_object_type_all   | not null
 disabled    | enum_boolean           | not null default 'n'::enum_boolean
 is_global   | enum_boolean           | not null default 'n'::enum_boolean
 uuid        | bytea                  |
Indexes:
    "icinga_zone_pkey" PRIMARY KEY, btree (id)
    "icinga_zone_object_name_key" UNIQUE CONSTRAINT, btree (object_name)
    "icinga_zone_uuid_key" UNIQUE CONSTRAINT, btree (uuid)
    "zone_parent" btree (parent_id)
Check constraints:
    "icinga_zone_uuid_check" CHECK (length(uuid) = 16)
Foreign-key constraints:
    "icinga_zone_parent_zone" FOREIGN KEY (parent_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE RESTRICT
Referenced by:
    TABLE "icinga_command" CONSTRAINT "icinga_command_zone" FOREIGN KEY (zone_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE RESTRICT
    TABLE "icinga_dependency" CONSTRAINT "icinga_dependency_zone" FOREIGN KEY (zone_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE RESTRICT
    TABLE "icinga_endpoint" CONSTRAINT "icinga_endpoint_zone" FOREIGN KEY (zone_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE RESTRICT
    TABLE "icinga_host" CONSTRAINT "icinga_host_zone" FOREIGN KEY (zone_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE RESTRICT
    TABLE "icinga_notification" CONSTRAINT "icinga_notification_zone" FOREIGN KEY (zone_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE RESTRICT
    TABLE "icinga_scheduled_downtime" CONSTRAINT "icinga_scheduled_downtime_zone" FOREIGN KEY (zone_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE RESTRICT
    TABLE "icinga_service" CONSTRAINT "icinga_service_zone" FOREIGN KEY (zone_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE RESTRICT
    TABLE "icinga_timeperiod" CONSTRAINT "icinga_timeperiod_zone" FOREIGN KEY (zone_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE RESTRICT
    TABLE "icinga_user" CONSTRAINT "icinga_user_zone" FOREIGN KEY (zone_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE RESTRICT
    TABLE "icinga_usergroup" CONSTRAINT "icinga_usergroup_zone" FOREIGN KEY (zone_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE RESTRICT
    TABLE "icinga_zone_inheritance" CONSTRAINT "icinga_zone_inheritance_parent_zone" FOREIGN KEY (parent_zone_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE RESTRICT
    TABLE "icinga_zone_inheritance" CONSTRAINT "icinga_zone_inheritance_zone" FOREIGN KEY (zone_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE CASCADE
    TABLE "icinga_zone" CONSTRAINT "icinga_zone_parent_zone" FOREIGN KEY (parent_id) REFERENCES icinga_zone(id) ON UPDATE CASCADE ON DELETE RESTRICT

director=> SELECT * FROM director_schema_migration;
 schema_version |        migration_time
----------------+-------------------------------
            167 | 2021-01-12 11:53:37.753429+01
            168 | 2021-03-17 10:52:29.085099+01
            169 | 2021-03-17 10:52:29.112534+01
            170 | 2021-03-17 10:52:29.118844+01
            171 | 2021-03-17 10:52:29.120196+01
            172 | 2022-02-25 14:18:38.933218+01
            173 | 2022-02-25 14:18:38.935847+01
(7 rows)

director=>
```

**Comment by maggu** [comment]
Hmm. pssql.sql for 1.8.1 clearly doesn't contain uuid for icinga_zone, and no migration before 174.

I don't see how I could have screwed it up, but maybe there was an initial error somewhere I missed. Might have been too hasty in reporting this, so maybe the issue should just be closed and I'll handle this issue manually.

**Comment by Thomas-Gelf** [comment]
Please run this, but step by step:

```sql
DROP INDEX icinga_zone_uuid_key;
ALTER TABLE icinga_zone ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX zone_uuid ON icinga_zone (uuid);

ALTER TABLE icinga_timeperiod ADD COLUMN uuid bytea CHECK(LENGTH(uuid) = 16);
UPDATE icinga_timeperiod SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_timeperiod ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX timeperiod_uuid ON icinga_timeperiod (uuid);

ALTER TABLE icinga_command ADD COLUMN uuid bytea CHECK(LENGTH(uuid) = 16);
UPDATE icinga_command SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_command ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX command_uuid ON icinga_command (uuid);

ALTER TABLE icinga_apiuser ADD COLUMN uuid bytea CHECK(LENGTH(uuid) = 16);
UPDATE icinga_apiuser SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_apiuser ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX apiuser_uuid ON icinga_apiuser (uuid);

ALTER TABLE icinga_endpoint ADD COLUMN uuid bytea CHECK(LENGTH(uuid) = 16);
UPDATE icinga_endpoint SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_endpoint ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX endpoint_uuid ON icinga_endpoint (uuid);

ALTER TABLE icinga_host ADD COLUMN uuid bytea CHECK(LENGTH(uuid) = 16);
UPDATE icinga_host SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_host ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX host_uuid ON icinga_host (uuid);

ALTER TABLE icinga_service ADD COLUMN uuid bytea CHECK(LENGTH(uuid) = 16);
UPDATE icinga_service SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_service ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX service_uuid ON icinga_service (uuid);

ALTER TABLE icinga_hostgroup ADD COLUMN uuid bytea CHECK(LENGTH(uuid) = 16);
UPDATE icinga_hostgroup SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_hostgroup ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX hostgroup_uuid ON icinga_hostgroup (uuid);

ALTER TABLE icinga_servicegroup ADD COLUMN uuid bytea CHECK(LENGTH(uuid) = 16);
UPDATE icinga_servicegroup SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_servicegroup ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX servicegroup_uuid ON icinga_servicegroup (uuid);

ALTER TABLE icinga_user ADD COLUMN uuid bytea CHECK(LENGTH(uuid) = 16);
UPDATE icinga_user SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_user ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX user_uuid ON icinga_user (uuid);

ALTER TABLE icinga_usergroup ADD COLUMN uuid bytea CHECK(LENGTH(uuid) = 16);
UPDATE icinga_usergroup SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_usergroup ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX usergroup_uuid ON icinga_usergroup (uuid);

ALTER TABLE icinga_notification ADD COLUMN uuid bytea CHECK(LENGTH(uuid) = 16);
UPDATE icinga_notification SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_notification ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX notification_uuid ON icinga_notification (uuid);

ALTER TABLE icinga_dependency ADD COLUMN uuid bytea CHECK(LENGTH(uuid) = 16);
UPDATE icinga_dependency SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_dependency ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX dependency_uuid ON icinga_dependency (uuid);

ALTER TABLE icinga_scheduled_downtime ADD COLUMN uuid bytea CHECK(LENGTH(uuid) = 16);
UPDATE icinga_scheduled_downtime SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_scheduled_downtime ALTER COLUMN uuid SET NOT NULL;
CREATE UNIQUE INDEX scheduled_downtime_uuid ON icinga_scheduled_downtime (uuid);

INSERT INTO director_schema_migration
(schema_version, migration_time)
VALUES (174, NOW());
```

In case one step fails please stop and let me know!

**Comment by Thomas-Gelf** [comment]
And don't worry, it's not your fault. I guess the combination of `ADD COLUMN uuid bytea **UNIQUE**` and `CREATE UNIQUE INDEX zone_uuid` is to blame here. The error message was probably slightly different at the first attempt.

**Comment by Thomas-Gelf** [comment]
I checked the PostgreSQL documentation, to figure out what the best approach would be. It says:

> Note: The preferred way to add a unique constraint to a table is ALTER TABLE ... ADD CONSTRAINT. The use of indexes to enforce unique constraints could be considered an implementation detail that should not be accessed directly. One should, however, be aware that there's no need to manually create indexes on unique columns; doing so would just duplicate the automatically-created index.

Please give me a second, we'll fix this differently.

**Comment by Thomas-Gelf** [comment]
Please try this one:

```sql
ALTER TABLE icinga_zone DROP COLUMN IF EXISTS uuidaa;

ALTER TABLE icinga_zone ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_zone SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_zone ALTER COLUMN uuid SET NOT NULL;

ALTER TABLE icinga_timeperiod ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_timeperiod SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_timeperiod ALTER COLUMN uuid SET NOT NULL;

ALTER TABLE icinga_command ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_command SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_command ALTER COLUMN uuid SET NOT NULL;

ALTER TABLE icinga_apiuser ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_apiuser SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_apiuser ALTER COLUMN uuid SET NOT NULL;

ALTER TABLE icinga_endpoint ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_endpoint SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_endpoint ALTER COLUMN uuid SET NOT NULL;

ALTER TABLE icinga_host ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_host SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_host ALTER COLUMN uuid SET NOT NULL;

ALTER TABLE icinga_service ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_service SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_service ALTER COLUMN uuid SET NOT NULL;

ALTER TABLE icinga_hostgroup ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_hostgroup SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_hostgroup ALTER COLUMN uuid SET NOT NULL;

ALTER TABLE icinga_servicegroup ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_servicegroup SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_servicegroup ALTER COLUMN uuid SET NOT NULL;

ALTER TABLE icinga_user ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_user SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_user ALTER COLUMN uuid SET NOT NULL;

ALTER TABLE icinga_usergroup ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_usergroup SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_usergroup ALTER COLUMN uuid SET NOT NULL;

ALTER TABLE icinga_notification ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_notification SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_notification ALTER COLUMN uuid SET NOT NULL;

ALTER TABLE icinga_dependency ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_dependency SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_dependency ALTER COLUMN uuid SET NOT NULL;

ALTER TABLE icinga_scheduled_downtime ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
UPDATE icinga_scheduled_downtime SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ALTER TABLE icinga_scheduled_downtime ALTER COLUMN uuid SET NOT NULL;

INSERT INTO director_schema_migration
(schema_version, migration_time)
VALUES (174, NOW());
```

Once again, in case one step fails please stop.

**Comment by maggu** [comment]
```
director=> ALTER TABLE icinga_zone DROP COLUMN IF EXISTS uuid;
ALTER TABLE
director=> ALTER TABLE icinga_zone ADD COLUMN uuid bytea UNIQUE CHECK(LENGTH(uuid) = 16);
NOTICE:  ALTER TABLE / ADD UNIQUE will create implicit index "icinga_zone_uuid_key" for table "icinga_zone"
ALTER TABLE
director=> UPDATE icinga_zone SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ERROR:  function gen_random_uuid() does not exist
LINE 1: UPDATE icinga_zone SET uuid = decode(replace(gen_random_uuid...
                                                     ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
director=>
```

**Comment by Thomas-Gelf** [comment]
Do you remember whether you skipped the `CREATE EXTENSION pgcrypto;` step at installation time? In case you did, please run this now. I have to fix the documentation, where it states that this is optional: it no longer is.

**Comment by Thomas-Gelf** [comment]
Afterwards please run all statements again, including the "DROP" one

**Comment by maggu** [comment]
Well. We do the entire installation and setup using Puppet, and it appears to already be there.

```
director=> SELECT * FROM pg_extension;
 extname  | extowner | extnamespace | extrelocatable | extversion | extconfig | extcondition
----------+----------+--------------+----------------+------------+-----------+--------------
 plpgsql  |       10 |           11 | f              | 1.0        |           |
 pgcrypto |       10 |         2200 | t              | 1.0        |           |
(2 rows)

director=> CREATE EXTENSION pgcrypto;
ERROR:  extension "pgcrypto" already exists
director=> UPDATE icinga_zone SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ERROR:  function gen_random_uuid() does not exist
LINE 1: UPDATE icinga_zone SET uuid = decode(replace(gen_random_uuid...
                                                     ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
director=> 

```

**Comment by maggu** [comment]
And according to this, the function exists from PostgreSQL 13.

https://www.postgresql.org/docs/13/functions-uuid.html

**Comment by Thomas-Gelf** [comment]
Oh, I see - you're on 9.2 :see_no_evil: It's in `pgcrypto` since [9.4](https://www.postgresql.org/docs/9.4/release-9-4.html)

**Comment by maggu** [comment]
Ah. Right. Well, I guess that was the initial issue anyway.

**Comment by maggu** [comment]
Thanks!

Well. We need a fix for #2256 pretty badly (with #2304 as a nice-to-have), so it will have to be upgrading PostgreSQL. Seems to be no problem doing that though. I don't even have to redeploy on RHEL 8.

**Comment by Thomas-Gelf** [comment]
@fl0wx: what PostgreSQL version are you running on your system?

**Comment by fl0wx** [comment]
@Thomas-Gelf PostgreSQL 11.14

**Comment by Thomas-Gelf** [comment]
Thank you! Did the migration proposed [here](https://github.com/Icinga/icingaweb2-module-director/issues/2482#issuecomment-1050910985) work for you?

**Comment by fl0wx** [comment]
@Thomas-Gelf didn't tried because I used a backup to restore my previous db state. When should I run that commands? Before or after the migrations in the UI? I have to start a new update if I should try this

**Comment by Thomas-Gelf** [comment]
It would be great, if you could take a new backup and then give v1.9 a new shot:

* run the upgrade, trigger migrations
* they'll fail, run the second block of statements [here](https://github.com/Icinga/icingaweb2-module-director/issues/2482#issuecomment-1050910985) manually on your Director database
* preferably do so step by step, in case one step fails please stop and let me know
* in case it works fine, you can then trigger the remaining migrations from the web UI

I prepared a fix including this, but before releasing it I'd prefer having it tested ;-)

**Comment by rafaelorafaelo** [comment]
Just to mention it: PostgreSQL 9.2 is the default version in RHEL/CentOS 7. It requires additional install/upgrade steps, if you raise the minimal version.

**Comment by DenisShuli** [comment]
Hello I did backup and tried new version but had the following errors.

Applying schema migrations in the GUI, after upgrading 1.8.1 to 1.9.0, results in the following error:

Migration 172 failed (SQLSTATE[42710]: Duplicate object: 7 ERROR: type "enum_sync_rule_purge_action" already exists) while running CREATE TYPE enum_sync_rule_purge_action AS ENUM('delete', 'disable') (Migration.php:60)

Then i run the second block of statements , but i had the following.

UPDATE icinga_zone SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ERROR:  function gen_random_uuid() does not exist
LINE 1: UPDATE icinga_zone SET uuid = decode(replace(gen_random_uuid...

director=# SELECT * FROM pg_extension;
 extname | extowner | extnamespace | extrelocatable | extversion | extconfig | extcondition
---------+----------+--------------+----------------+------------+-----------+--------------
 plpgsql |       10 |           11 | f              | 1.0        |           |
(1 row)

Maybe i have to install pgcrypto extension? 


Your Environment
Director version (System - About): 1.9.0
Icinga Web 2 version and modules (System - About): 2.9.5
Icinga 2 version (icinga2 --version): 2.13.2-1
Operating System and version: CentOS Linux release 7.9.2009 (Core)
Webserver, PHP versions: Apache HTTP Server 2.4.6, PHP 7.3.29
Database: PostgreSQL server 9.6

**Comment by DenisShuli** [comment]
Hello I did backup and tried new version but had the following errors.

Applying schema migrations in the GUI, after upgrading 1.8.1 to 1.9.0, results in the following error:

Migration 172 failed (SQLSTATE[42710]: Duplicate object: 7 ERROR: type "enum_sync_rule_purge_action" already exists) while running CREATE TYPE enum_sync_rule_purge_action AS ENUM('delete', 'disable') (Migration.php:60)

Then i run the second block of statements , but i had the following.

UPDATE icinga_zone SET uuid = decode(replace(gen_random_uuid()::text, '-', ''), 'hex') WHERE uuid IS NULL;
ERROR: function gen_random_uuid() does not exist
LINE 1: UPDATE icinga_zone SET uuid = decode(replace(gen_random_uuid...

director=# SELECT * FROM pg_extension;
extname | extowner | extnamespace | extrelocatable | extversion | extconfig | extcondition
---------+----------+--------------+----------------+------------+-----------+--------------
plpgsql | 10 | 11 | f | 1.0 | |
(1 row)

Maybe i have to install pgcrypto extension?

**Comment by rafaelorafaelo** [comment]
> Maybe i have to install pgcrypto extension?

Hi, gen_random_uuid() is part of the pgcrypto extension https://www.postgresql.org/docs/14/pgcrypto.html#id-1.11.7.35.10 You should install the extension, usually some kind of contrib package, and create it with CREATE EXTENSION https://www.postgresql.org/docs/14/sql-createextension.html

I would suggest to wait for the next release and go back to 1.8.1. This function doesn't exist in the minimal recommended PostgreSQL version (PostgreSQL (>= 9.1)). There is no final solution for this problem at all.

**Comment by fl0wx** [comment]
@Thomas-Gelf  sorry couldnt managed to try this again for now, but i found out, that our pgsql Database is missing the uuid method. As soon as we upgraded the PostgreSQL Version (method is included in 13+ default) ill try the update again. Also there should be a note to the dependency of the uuid method (pgsql13+ or pgcrypto extension). Currently it is stated as optional

**Comment by bahrpatrick** [comment]
For me it helped to run "CREATE EXTENSION pgcrypto;" in the database before doing the upgrade.
After this all the migration scripts completed successfully.
Maybe you should mention this in the upgrade / installation documentation or add it to the migration scripts.

**Comment by Yoda-BZH** [comment]
Hello,

The same errors exists for mysql, see #2492

**Comment by Thomas-Gelf** [comment]
The initial issue has been fixed, #2520 has been created for pgsql v9.2, #2492 is not related and will be addressed separately.

**Comment by viper539** [comment]
> Do you remember whether you skipped the `CREATE EXTENSION pgcrypto;` step at installation time? In case you did, please run this now. I have to fix the documentation, where it states that this is optional: it no longer is.

This fixed it for me.
Upgrade from 1.7.1 to 1.9.1

Thank you!
